### PR TITLE
cmd/protoc-gen-go-grpc: add hooks to allow overriding generated code

### PR DESCRIPTION
**Pull request by ZhouyihaiDing**
This PR adds a hook which users can implement and insert helper codes to the codegen, eg service renaming.

RELEASE NOTES: N/A

### CODE REVIEW

**Comment by easwars** [comment]
Could you please add a PR description explaining what this PR attempts to do. Thanks.

File: cmd/protoc-gen-go-grpc/grpc.go [hunk 1]
```diff
@@ -121,6 +144,9 @@ func genService(gen *protogen.Plugin, file *protogen.File, g *protogen.Generated
 		g.P(deprecationComment)
 	}
 	g.P("func New", clientName, " (cc ", grpcPackage.Ident("ClientConnInterface"), ") ", clientName, " {")
+	if helper != nil {
```
Comment (dfawley) [inline cmd/protoc-gen-go-grpc/grpc.go:147]: Can the helper be always set, with a default implementation which does the things that we do when `helper == nil`?

Comment (ZhouyihaiDing) [reply cmd/protoc-gen-go-grpc/grpc.go]: Done. Removed check.

**Comment by ZhouyihaiDing** [comment]
Hi Doug,
Can you please help take a look at this PR again when you have time.
Thanks!

**Comment by ZhouyihaiDing** [comment]
We need to modify the body of the handler.
eg:
From
```
info := &grpc.UnaryServerInfo{
		Server:     srv,
		FullMethod: "/" + "grpc.testing.Test" + "/UnaryCall",
	}
```
to 
```
info := &grpc.UnaryServerInfo{
		Server:     srv,
		FullMethod: "/" + TestHelper.processServiceName("grpc.testing.Test") + "/UnaryCall",
	}
```

FullMethod name is no longer a hard coded string. We need to replace it with a map lookup.

**Comment by menghanl** [comment]
And because the function body needs to be changed, the idea of a separate standalone code generator doesn't work.

File: cmd/protoc-gen-go-grpc/grpc.go [hunk 2]
```diff
@@ -34,6 +34,52 @@ const (
...
+	if *requireUnimplemented {
+		g.P("func (Unimplemented", serverType, ") mustEmbedUnimplemented", serverType, "() {}")
+	}
+	g.P()
+}
+
+func (serviceGenerateHelper) generateServerSideHelperCodes(gen *protogen.Plugin, file *protogen.File, g *protogen.GeneratedFile, service *protogen.Service, serverType string, serviceDescVar string) {
+}
+
+type serviceGenerateHelperImpl struct {
```
Comment (dfawley) [inline cmd/protoc-gen-go-grpc/grpc.go:77]: This type seems unnecessary.

File: cmd/protoc-gen-go-grpc/grpc.go [hunk 3]
```diff
@@ -121,6 +167,7 @@ func genService(gen *protogen.Plugin, file *protogen.File, g *protogen.Generated
 		g.P(deprecationComment)
 	}
 	g.P("func New", clientName, " (cc ", grpcPackage.Ident("ClientConnInterface"), ") ", clientName, " {")
+	helper.preprocessServiceName(g, service, clientName)
```
Comment (dfawley) [inline cmd/protoc-gen-go-grpc/grpc.go:170]: How would this be used?  Maybe name it `generateNewClientDefinitions` or something?

File: cmd/protoc-gen-go-grpc/grpc.go [hunk 4]
```diff
@@ -34,6 +34,52 @@ const (
 	statusPackage  = protogen.GoImportPath("google.golang.org/grpc/status")
 )
 
+type serviceGenerateHelper struct {
+	properties map[string]string
```
Comment (dfawley) [inline cmd/protoc-gen-go-grpc/grpc.go:38]: What is this map used for?

File: cmd/protoc-gen-go-grpc/grpc.go [hunk 5]
```diff
@@ -170,23 +217,7 @@ func genService(gen *protogen.Plugin, file *protogen.File, g *protogen.Generated
...
-		}
-		g.P("func (Unimplemented", serverType, ") ", serverSignature(g, method), "{")
-		g.P("return ", nilArg, statusPackage.Ident("Errorf"), "(", codesPackage.Ident("Unimplemented"), `, "method `, method.GoName, ` not implemented")`)
-		g.P("}")
-	}
-	if *requireUnimplemented {
-		g.P("func (Unimplemented", serverType, ") mustEmbedUnimplemented", serverType, "() {}")
-	}
-	g.P()
+	helper.generateUmplementedHelperCodes(gen, file, g, service)
```
Comment (dfawley) [inline cmd/protoc-gen-go-grpc/grpc.go:220]: `generateUnimplementedServerType`?

File: cmd/protoc-gen-go-grpc/grpc.go [hunk 6]
```diff
@@ -206,6 +237,8 @@ func genService(gen *protogen.Plugin, file *protogen.File, g *protogen.Generated
 	g.P("}")
 	g.P()
 
+	helper.generateServerSideHelperCodes(gen, file, g, service, serverType, serviceDescVar)
```
Comment (dfawley) [inline cmd/protoc-gen-go-grpc/grpc.go:240]: `generateServerFunctions`?

File: cmd/protoc-gen-go-grpc/grpc.go [hunk 7]
```diff
@@ -270,7 +303,8 @@ func clientSignature(g *protogen.GeneratedFile, method *protogen.Method) string
 
 func genClientMethod(gen *protogen.Plugin, file *protogen.File, g *protogen.GeneratedFile, method *protogen.Method, index int) {
 	service := method.Parent
-	sname := fmt.Sprintf("/%s/%s", service.Desc.FullName(), method.Desc.Name())
+	serviceName := helper.generateProcessedServiceName(service)
```
Comment (dfawley) [inline cmd/protoc-gen-go-grpc/grpc.go:306]: There's no code generation happening here.  Maybe `formatServiceName`?

File: cmd/protoc-gen-go-grpc/grpc.go (hunk 3 above)
Comment (ZhouyihaiDing) [reply cmd/protoc-gen-go-grpc/grpc.go:170]: Done

File: cmd/protoc-gen-go-grpc/grpc.go (hunk 6 above)
Comment (ZhouyihaiDing) [reply cmd/protoc-gen-go-grpc/grpc.go]: Done

File: cmd/protoc-gen-go-grpc/grpc.go (hunk 7 above)
Comment (ZhouyihaiDing) [reply cmd/protoc-gen-go-grpc/grpc.go]: Done

File: cmd/protoc-gen-go-grpc/grpc.go (hunk 5 above)
Comment (ZhouyihaiDing) [reply cmd/protoc-gen-go-grpc/grpc.go]: Done

File: cmd/protoc-gen-go-grpc/grpc.go (hunk 2 above)
Comment (ZhouyihaiDing) [reply cmd/protoc-gen-go-grpc/grpc.go:77]: If `helper` is `serviceGenerateHelper`, then I am not allowed to override it's method internally.
eg: `method generateNewClientDefinitions already declared for type serviceGenerateHelper`

Thus I need a `serviceGenerateHelperImpl` so I can override it's methods internally.

Comment (dfawley) [reply cmd/protoc-gen-go-grpc/grpc.go:77]: Sounds like you want an `interface` instead?

File: cmd/protoc-gen-go-grpc/grpc.go (hunk 4 above)
Comment (ZhouyihaiDing) [reply cmd/protoc-gen-go-grpc/grpc.go:38]: It records if the user enables renaming or not.

I record this information internally when create/override the helper:
`helper = &serviceGenerateHelperImpl{serviceGenerateHelper{properties: map[string]string{"rename": "false"}}}`

Using a map may look more generic than using a bool.

Comment (dfawley) [reply cmd/protoc-gen-go-grpc/grpc.go:38]: You should be able to keep those implementation details out of OSS, I think.

Comment (ZhouyihaiDing) [reply cmd/protoc-gen-go-grpc/grpc.go]: Done. Removed

File: cmd/protoc-gen-go-grpc/grpc.go (hunk 2 above)
Comment (ZhouyihaiDing) [reply cmd/protoc-gen-go-grpc/grpc.go]: Done. Changed it to interface

File: cmd/protoc-gen-go-grpc/grpc.go [hunk 8]
```diff
@@ -34,6 +34,55 @@ const (
...
 
+type serviceGenerateHelperInterface interface {
+	formatServiceName(service *protogen.Service) string
+	generateNewClientDefinitions(g *protogen.GeneratedFile, service *protogen.Service, clientName string)
+	generateUnimplementedServerType(gen *protogen.Plugin, file *protogen.File, g *protogen.GeneratedFile, service *protogen.Service)
+	generateServerFunctions(gen *protogen.Plugin, file *protogen.File, g *protogen.GeneratedFile, service *protogen.Service, serverType string, serviceDescVar string)
+}
+
+type serviceGenerateHelper struct {
+	serviceGenerateHelperInterface
```
Comment (dfawley) [inline cmd/protoc-gen-go-grpc/grpc.go:45]: No need to embed this.

File: cmd/protoc-gen-go-grpc/grpc.go [hunk 9]
```diff
@@ -34,6 +34,55 @@ const (
...
+	if *requireUnimplemented {
+		g.P("func (Unimplemented", serverType, ") mustEmbedUnimplemented", serverType, "() {}")
+	}
+	g.P()
+}
+
+func (serviceGenerateHelper) generateServerFunctions(gen *protogen.Plugin, file *protogen.File, g *protogen.GeneratedFile, service *protogen.Service, serverType string, serviceDescVar string) {
+}
+
+var helper serviceGenerateHelperInterface = new(serviceGenerateHelper)
```
Comment (dfawley) [inline cmd/protoc-gen-go-grpc/grpc.go:84]: `var helper serviceGenerateHelperInterface = serviceGenerateHelper{}`

**Review by dfawley — APPROVED** [review]
Looks good; just a couple nits.

For the future, please don't overwrite commits and force push.  Instead add a new commit and do a normal push.  The one time we accept force pushes is if you need to rebase your branch from master.  Having a separate commit for each round of changes is useful to see what was changed since the last round of comments.

File: cmd/protoc-gen-go-grpc/grpc.go (hunk 9 above)
Comment (ZhouyihaiDing) [reply cmd/protoc-gen-go-grpc/grpc.go]: Done.

File: cmd/protoc-gen-go-grpc/grpc.go [hunk 10]
```diff
@@ -374,7 +409,8 @@ func genServerMethod(gen *protogen.Plugin, file *protogen.File, g *protogen.Gene
 		g.P("if interceptor == nil { return srv.(", service.GoName, "Server).", method.GoName, "(ctx, in) }")
 		g.P("info := &", grpcPackage.Ident("UnaryServerInfo"), "{")
 		g.P("Server: srv,")
-		g.P("FullMethod: ", strconv.Quote(fmt.Sprintf("/%s/%s", service.Desc.FullName(), method.Desc.Name())), ",")
+		serviceName := helper.formatServiceName(service)
```
Comment (menghanl) [inline cmd/protoc-gen-go-grpc/grpc.go:412]: Hmm, what if the method is also renamed? Is that never changed?
Maybe make this take service and method, and return full name.

**Review by menghanl — APPROVED** [review]

**Comment by menghanl** [comment]
Rebase on master

**Review by dfawley — APPROVED** [review]

**Pull request merged**